config/import-control.xml

```diff
@@ -165,6 +165,13 @@
     <allow class="com.puppycrawl.tools.checkstyle.JavadocDetailNodeParser" local-only="true"/>
     <allow class="com.google.common.collect.ImmutableList" local-only="true"/>
     <allow class="java.nio.charset.StandardCharsets" local-only="true" />
+    <disallow pkg="com\.puppycrawl\.tools\.checkstyle\.(checks|ant|doclets|filters)" regex="true"/>
+
+    <class name=".*Model" regex="true">
```

Comment:
What about *Presentation classes ?

Reply:
The example was empty.

> `<!-- almost all should be allowed -->`

If we move all current allows to inside the class, we get 79 violations because not all classes are handled (BaseCellEditor, CodeSelector, etc...) and the parent doesn't have allows for them either since your example removed ` strategyOnMismatch="disallowed"`.
If we are not moving, but duplicating the allows, then `*Presentation` serves no purpose as it will have the same allows.

Reply:
> *Presentation classes 

I missed that you said presentation and not panel/view/dialog. Your example at the bottom of https://github.com/checkstyle/checkstyle/issues/3492#issuecomment-376301219 had nothing with presentation.

Reply:
It better to add it. I not aware that there are some code standard on naming of such classes, but usually it is smth like about Model and Presentation Model(https://martinfowler.com/eaaDev/PresentationModel.html) (*PresentationModel, *Model, *PM, *Pm). Such classes suppose to keep whole logic and UI class delegate calls to it. So logic could be UTed and UI part requires less attention as there is minimal logic in it.

Reply:
@romani Please specify what imports you want me allow/disallow for `*PresentationModel, *Model, *PM, *Pm`. Is it the exact same as https://github.com/checkstyle/checkstyle/pull/5641#discussion_r177435751 .
What about `*(Panel|View|Dialog)`?

Reply:
> Please specify what imports you want me allow/disallow for *PresentationModel, *Model, *PM, *Pm. Is it the exact same as #5641 (comment) ?

yes, disallow "java.swing.J*" (all visual/rendering classes)

> What about *(Panel|View|Dialog)?

Ideally I would move enforce such classes  to use Presentation classes and classes from jdk and nothing else. So to force all rendering preparation(even toString) to be delegated to Presentation. But this is completely different issue.
To make atleast some step in front of this goal:
We can allow from checkstyle only `import com.puppycrawl.tools.checkstyle.api.DetailAST;`
and `import com.puppycrawl.tools.checkstyle.api.DetailNode;`. All other should be delegated to Presentation class.
If that is too much to do now .... my preference is to use "blacklist" of imports here. So we allow all, except for some known odd imports (from "checks", from "ant", ... )

Reply:
> > Please specify what imports you want me allow/disallow for *PresentationModel, *Model, *PM, *Pm. Is it the exact same as #5641 (comment) ?

> yes, disallow "java.swing.J*" (all visual/rendering classes)

This is already being done in `.*Model` which includes `.*PresentationModel`.

> > What about *(Panel|View|Dialog)?

> We can allow from checkstyle only import com.puppycrawl.tools.checkstyle.api.DetailAST;
and import com.puppycrawl.tools.checkstyle.api.DetailNode;. All other should be delegated to Presentation class.

Moving `allow pkg="com.puppycrawl.tools.checkstyle.api"` from whole package to class results in 12 violations. Other classes use these imports: CodeSelector, CodeSelectorPresentation, MainFrame, MainFrameModel, ParseTreeTableModel, ParseTreeTablePresentation

> So we allow all, except for some known odd imports (from "checks", from "ant", ... )

We already have this on the package level.
Do you want me to move this to the class? There is no point duplicating it for the class when the package already does this.

Reply:
please update to enforce no "swing.J*" at "*Presentation" classes.

StandardCharsets, JavadocDetailNodeParser, JavaParser, checkstyle.utils - should be allowed only in Models&Presentations . There should not be problem, as they are in use only in presentation models.

generally in UI it is allowed to use only (from whole "api"):
```
import com.puppycrawl.tools.checkstyle.api.DetailAST;
import com.puppycrawl.tools.checkstyle.api.DetailNode;
import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
```
I do not see any other usages in code and this is correct.
you can create separate issue for refactoring of UI to move even this imports to model only.

Reply:
> no "swing.J*" at "*Presentation" classes
> StandardCharsets, JavadocDetailNodeParser, JavaParser, checkstyle.utils - should be allowed only in Models&Presentations 

Done.

> you can create separate issue for refactoring of UI to move even this imports to model only.

You would need to create such issue as I don't know what imports or classes you are referring to.

Reply:
new issue is opened - #5697